cetmix_tower_server/models/cx_tower_vault_mixin.py

```diff
@@ -0,0 +1,324 @@
+        for record in self:
+            for field, value in vals.items():
+                # Find existing vault record using filtered (in-memory operation)
+                existing_record = existing_vault_records.filtered(
+                    lambda v, record_id=record.id, field_name=field: (
+                        v.res_id == record_id and v.field_name == field_name
+                    )
+                )
+
+                if existing_record:
+                    if value:
+                        vals_to_update_records[value] |= existing_record
+
+                    else:
+                        records_to_unlink |= existing_record
+
+                else:
+                    if not value:
+                        continue
+
+                    records_to_create.append(
+                        {
+                            "res_model": self._name,
+                            "res_id": record.id,
+                            "field_name": field,
+                            "data": value,
+                        }
+                    )
+
+        # Batch operations
+        for value, records in vals_to_update_records.items():
+            records.sudo().write({"data": value})
+
+        self.env["cx.tower.vault"].sudo().create(records_to_create)
+        records_to_unlink.sudo().unlink()
+
+    @api.model_create_multi
+    def create(self, vals_list):
+        """Create records and set secret values if present"""
+
+        # Step 1: Extract secret fields and generate temporary IDs
+        secret_vals, modified_vals_list = self._extract_secret_fields(vals_list)
+
+        # Step 2: Create records with batch operation
+        records = super().create(modified_vals_list)
+
+        # Step 3: Update vault records with real IDs
+        if secret_vals:
+            self._process_secret_values_after_creation(records, secret_vals)
+
+        return records
+
+    def _extract_secret_fields(self, vals_list):
+        """Extract secret fields and replace with temporary identifiers.
+
+        Returns:
+            tuple: (secret_vals dict, modified vals_list)
+        """
+        temp_id_counter = 0
+        secret_vals = {}
+
+        for vals in vals_list:
+            for secret_field in self.SECRET_FIELDS:
+                if secret_field in vals and vals[secret_field]:
+                    temp_id_counter += 1
+                    temp_identifier = f"temp_secret_{temp_id_counter}"
+                    secret_vals[temp_identifier] = vals[secret_field]
+                    vals[secret_field] = temp_identifier
+
+        return secret_vals, vals_list
+
+    def _process_secret_values_after_creation(self, records, secret_vals):
+        """Process secret values after records creation."""
+        records_dict = self._fetch_records_data(records)
+
+        for record_dict in records_dict:
+            self._process_single_record_secrets(record_dict, secret_vals)
+
+    def _fetch_records_data(self, records):
```

Comment:
This can be put inside _process_secret_values_after_creation as it doesn't contain any specific business logic.